src/library/scala/collection/Iterator.scala

```diff
@@ -957,6 +958,37 @@ object Iterator extends IterableFactory[Iterator] {
     def next() = elem
   }
 
+  /**
+    * @return an Iterator by using a function `f` producing elements of
+    *         type `A` and updating an internal state `S`.
+    * @param init State initial value
+    * @param f    Computes the next element (or returns `None` to signal
+    *             the end of the collection)
+    * @tparam A   Type of the elements
+    * @tparam S   Type of the internal state
+    */
+  def unfold[A, S](init: S)(f: S => Option[(A, S)]): Iterator[A] = new AbstractIterator[A] {
+    private[this] var state: S = init
+    private[this] var nextResult: Option[(A, S)] = null
+
+    override def hasNext: Boolean = {
+      if (nextResult eq null) {
+        nextResult = Objects.requireNonNull(f(state), "null during unfold")
```

Comment:
This logic isn't safe if `f` is side-effecting--for instance, if it generates a random sequence that terminates.  You need to have a separate var that remembers when you've hit the last item or somesuch.  Personally I'd encode it as a `uninitialized` bool, and just put `(A, S) ` in its own var with it being `null` as the indication that the stream has stopped (and then pack `init` into `(null.asInstanceOf[A], init)` to get going).

Reply:
Or better yet use two fields, one for `S` and one for `A`, and either bools or integer state to keep track of what's going on.